src/DNS_Mgr.h

```diff
@@ -131,12 +133,10 @@ class DNS_Mgr : public iosource::IOSource {
 	void DoProcess();
 
 	// IOSource interface.
-	void GetFds(iosource::FD_Set* read, iosource::FD_Set* write,
-	                    iosource::FD_Set* except) override;
-	double NextTimestamp(double* network_time) override;
 	void Process() override;
-	void Init() override;
-	const char* Tag() override { return "DNS_Mgr"; }
+	void InitSource() override;
+	const char* Tag() override	{ return "DNS_Mgr"; }
+	double GetNextTimeout() override	{ return -1; }
```

Comment:
I didn't, and you're right that the old code checked for whether `async_timeouts` had anything in it when returning from `NextTimestamp` and loops through them checking for timeouts in `DoProcess`.

I'm a little leery of adding a timer object for every async timeout though, as that seems like it could up with a ton of timers. Right now `DNS_Mgr` returns a -1 all the time. It could easily return a short timeout if it has something in the queue. `Process()` hasn't changed at all since the old code so it should process timeouts correctly.